File: templates/problems/06_linked_list/01_reverse_list/typescript/solution.ts
import { ListNode } from "./helpers";

export function solveReverseList(head: ListNode | null): ListNode | null {
    let prev: ListNode | null = null;
    let curr = head;
    while (curr !== null) {
        const next = curr.next;
        curr.next = prev;
        prev = curr;
        curr = next;
    }
    return prev;
}

File: templates/problems/06_linked_list/01_reverse_list/python/solution.py
from helpers import ListNode, Optional


def solve_reverse_list(head: Optional[ListNode]) -> Optional[ListNode]:
    """Reverse a singly linked list iteratively.

    Time: O(n), Space: O(1).
    """
    prev: Optional[ListNode] = None
    curr = head
    while curr is not None:
        nxt = curr.next
        curr.next = prev
        prev = curr
        curr = nxt
    return prev
